File: 2026-03-March-GFG-POTD/March 21 - Number of BST From Array/Solution.py
class Solution:
    def countBSTs(self, arr):
        n = len(arr)
        
        # Step 1: Precompute Catalan numbers
        catalan = [0] * (n + 1)
        catalan[0] = 1
        if n >= 1:
            catalan[1] = 1
        
        for i in range(2, n + 1):
            for j in range(i):
                catalan[i] += catalan[j] * catalan[i - j - 1]
        
        result = []
        
        # Step 2: Compute answer for each root
        for i in range(n):
            root = arr[i]
            leftCount = 0
            rightCount = 0
            
            for j in range(n):
                if arr[j] < root:
                    leftCount += 1
                elif arr[j] > root:
                    rightCount += 1
            
            result.append(catalan[leftCount] * catalan[rightCount])
        
        return result

File: 2026-03-March-GFG-POTD/March 21 - Number of BST From Array/Solution.cpp
class Solution {
public:
    vector<int> countBSTs(vector<int>& arr) {
        int n = arr.size();
        
        // Step 1: Precompute Catalan numbers
        vector<int> catalan(n + 1, 0);
        catalan[0] = 1;
        if (n >= 1) catalan[1] = 1;
        
        for (int i = 2; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                catalan[i] += catalan[j] * catalan[i - j - 1];
            }
        }
        
        vector<int> result;
        
        // Step 2: Compute answer for each root
        for (int i = 0; i < n; i++) {
            int root = arr[i];
            int leftCount = 0, rightCount = 0;
            
            for (int j = 0; j < n; j++) {
                if (arr[j] < root) leftCount++;
                else if (arr[j] > root) rightCount++;
            }
            
            result.push_back(catalan[leftCount] * catalan[rightCount]);
        }
        
        return result;
    }
};
